src/app_manager/app_manager.py

```diff
@@ -638,6 +639,15 @@ def is_done(target):
             if is_launch:
                 return (not target.pm or target.pm.done)
             return target.stopped
+        def check_required(target):
+            # required nodes are not registered to the dead_list when finished
+            # so we need to constantly check its return value
+            if is_launch and target.pm:
```

Comment:
I want to also check `rosrun` case, too.

Reply:
If a node in the rosrun case is registered as required, it will kill the whole default_launch when exiting. Since we don't have any code to restart the default_launch, this would mean that we won't be able to run the same app or any other rosrun app again.
Therefore, the rosrun should never have required nodes, and if it does we will have bigger problems than setting the exit code.

Nevertheless, I do agree that guarding for both cases is good practice.

Reply:
Oh, I see.